youtube_dl/extractor/xiami.py

```diff
@@ -0,0 +1,131 @@
+            }
+        },
+        {
+            'url': 'http://www.xiami.com/song/1775256504',
+            'md5': '932a3abd45c6aa2b1fdbe028fcb4c4fc',
+            'info_dict': {
+                'id': '1775256504',
+                'ext': 'mp3',
+                'title': '悟空',
+                'creator': '戴荃',
+                'album': '悟空',
+                'description': 'md5:206e67e84f9bed1d473d04196a00b990',
+            }
+        },
+        {
+            'url': 'http://www.xiami.com/album/2100300444',
+            'info_dict': {
+                'id': '2100300444',
+            },
+            'playlist_count': 10,
+        },
+        {
+            'url': 'http://www.xiami.com/album/512288?spm=a1z1s.6843761.1110925389.6.hhE9p9',
+            'only_matching': True,
+        }
+    ]
+
+    _TYPES = {
+        'song': '',
+        'album': '/type/1',
+        'artist': '/type/2',
+        'collect': '/type/3',
+    }
+
+    _XML_BASE_URL = 'http://www.xiami.com/song/playlist/id'
+    _NS_MAP = {'xm': 'http://xspf.org/ns/0/'}
+
+    def _real_extract(self, url):
+        match = re.match(self._VALID_URL, url)
+        assert(match)
+        typ, _id = match.group('typ'), match.group('id')
+
+        if typ not in self._TYPES:
+            raise ExtractorError('URL not supported')
+        xml_url = '%s/%s' % (self._XML_BASE_URL, _id)
+        xml_url += self._TYPES[typ]
+
+        playlist = self._download_xml(xml_url, _id)
+        tracklist = xpath_element(playlist, xpath_with_ns('./xm:trackList', self._NS_MAP))
+
+        def _extract_single(track):
+            artist = xpath_text(track, xpath_with_ns('xm:artist', self._NS_MAP), default='')
+            artist = artist.split(';')
+
+            ret = {
+                'id': xpath_text(track, xpath_with_ns('xm:song_id', self._NS_MAP)),
+                'title': xpath_text(track, xpath_with_ns('xm:title', self._NS_MAP)),
+                'album': xpath_text(track, xpath_with_ns('xm:album_name', self._NS_MAP)),
+                'artist': ';'.join(artist) if artist else None,
+                'creator': artist[0] if artist else None,
+                'url': self._decrypt(xpath_text(track, xpath_with_ns('xm:location', self._NS_MAP))),
+                'thumbnail': xpath_text(track, xpath_with_ns('xm:pic', self._NS_MAP), default=None),
+                'duration': int_or_none(xpath_text(track, xpath_with_ns('xm:length', self._NS_MAP))),
+            }
+
+            lyrics_url = xpath_text(track, xpath_with_ns('xm:lyric', self._NS_MAP))
+            if lyrics_url and lyrics_url.endswith('.lrc'):
+                ret['description'] = self._download_webpage(lyrics_url, ret['id'])
+            return ret
+
+        result = [_extract_single(track) for track in tracklist]
+        if not result:
+            raise ExtractorError('No track found')
+
+        if typ == 'song':
+            return result[0]
+        else:
+            return self.playlist_result(result, _id)
```

Comment:
Instead of this song and album scenarios should be split into separate extractors.